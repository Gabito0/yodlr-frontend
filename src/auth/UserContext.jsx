import React from "react";

/**
 * UserContext provides a context for user information in the application.
 *
 * This context allows different parts of the app to access and update the current
 * user's information, such as their ID, email, and role (e.g., admin status).
 * It facilitates state management for user data across various components.
 *
 * Components can consume this context to access user data or to make updates
 * when user information changes, such as during login, logout, or profile updates.
 */

const UserContext = React.createContext();

export default UserContext;
