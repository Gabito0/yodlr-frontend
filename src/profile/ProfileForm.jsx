import React, { useState, useContext } from "react";
import { useNavigate } from "react-router-dom";
import AppAlert from "../common/AppAlert";
import UserContext from "../auth/UserContext";
import userApi from "../api/api";
import LoadingSpinner from "../common/LoadingSpinner";

/**
 * ProfileForm component renders a form for editing user profile information.
 *
 * It manages the state of the form data and handles form submission.
 * Upon successful save, it updates the current user context and clears form errors.
 * It also provides functionality to delete the user account.
 */
const ProfileForm = ({ logout }) => {
  const navigate = useNavigate(); // Hook for navigation
  const { currentUser, setCurrentUser } = useContext(UserContext); // Get current user from context

  // State for form data, form errors, save confirmation, and loading
  const [formData, setFormData] = useState({
    email: currentUser.email,
    firstName: currentUser.firstName,
    lastName: currentUser.lastName,
    state: currentUser.state,
    currentPassword: "",
    newPassword: "",
  });
  const [formErrors, setFormErrors] = useState([]);
  const [saveConfirmed, setSaveConfirmed] = useState(false);
  const [loading, setLoading] = useState(false);

  // console.debug(
  //   "ProfileForm",
  //   "currentUser=",
  //   currentUser,
  //   "formData=",
  //   formData,
  //   "formErrors=",
  //   formErrors,
  //   "saveConfirmed=",
  //   saveConfirmed,
  //   "loading=",
  //   loading
  // );

  // Handle form submission
  async function handleSubmit(evt) {
    evt.preventDefault();
    setLoading(true);

    let profileData = { ...formData };
    if (!formData.newPassword) {
      delete profileData.newPassword;
    }

    try {
      const updatedUser = await userApi.saveProfile(
        currentUser.id,
        profileData
      );
      setCurrentUser(updatedUser);
      setFormErrors([]);
      setSaveConfirmed(true);
    } catch (errors) {
      setFormErrors(errors);
      setSaveConfirmed(false);
    } finally {
      setLoading(false);
    }

    // Clear password fields
    setFormData((f) => ({
      ...f,
      currentPassword: "",
      newPassword: "",
    }));
  }

  // Update form data state on input change
  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((f) => ({ ...f, [name]: value }));
    setFormErrors([]);
  }

  // Handle user account deletion
  async function handleDelete() {
    setLoading(true);
    try {
      await userApi.deleteUser(currentUser.id);
      logout(); // Ensure this function clears the session/token
      navigate("/login");
    } catch (err) {
      setFormErrors([err.message]);
    } finally {
      setLoading(false);
    }
  }

  return (
    <div className="container mt-5">
      <div className="row justify-content-center">
        <div className="col-md-8">
          <div className="card">
            <div className="card-header">
              <h3>Profile</h3>
            </div>
            <div className="card-body">
              {loading ? (
                <LoadingSpinner />
              ) : (
                <form onSubmit={handleSubmit}>
                  <div className="form-group">
                    <label htmlFor="email">Email</label>
                    <input
                      type="email"
                      id="email"
                      name="email"
                      className="form-control"
                      value={formData.email}
                      onChange={handleChange}
                      autoComplete="email"
                    />
                  </div>
                  <div className="form-group">
                    <label htmlFor="firstName">First Name</label>
                    <input
                      id="firstName"
                      name="firstName"
                      className="form-control"
                      value={formData.firstName}
                      onChange={handleChange}
                    />
                  </div>
                  <div className="form-group">
                    <label htmlFor="lastName">Last Name</label>
                    <input
                      id="lastName"
                      name="lastName"
                      className="form-control"
                      value={formData.lastName}
                      onChange={handleChange}
                    />
                  </div>
                  <div className="form-group">
                    <label htmlFor="currentPassword">Current Password</label>
                    <input
                      type="password"
                      id="currentPassword"
                      name="currentPassword"
                      className="form-control"
                      value={formData.currentPassword}
                      onChange={handleChange}
                      autoComplete="current-password"
                    />
                  </div>
                  <div className="form-group">
                    <label htmlFor="newPassword">New Password</label>
                    <input
                      type="password"
                      id="newPassword"
                      name="newPassword"
                      className="form-control"
                      value={formData.newPassword}
                      onChange={handleChange}
                      autoComplete="new-password"
                    />
                  </div>
                  <div className="form-group d-flex align-items-center">
                    <label htmlFor="state" className="mr-2">
                      State:
                    </label>
                    <p className="form-control-plaintext mb-0 ms-1" id="state">
                      {formData.state}
                    </p>
                  </div>

                  {formErrors.length > 0 && (
                    <AppAlert type="danger" messages={formErrors} />
                  )}

                  {saveConfirmed && (
                    <AppAlert
                      type="success"
                      messages={["Updated successfully."]}
                    />
                  )}

                  <div className="d-flex justify-content-between">
                    <button className="btn btn-primary" type="submit">
                      Save Changes
                    </button>
                    <button
                      className="btn btn-danger"
                      type="button"
                      onClick={handleDelete}
                    >
                      Delete Account
                    </button>
                  </div>
                </form>
              )}
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default ProfileForm;
