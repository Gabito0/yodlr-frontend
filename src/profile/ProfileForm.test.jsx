import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import ProfileForm from "./ProfileForm";
import UserContext from "../auth/UserContext";
import userApi from "../api/api";
import { vi } from "vitest";

vi.mock("../api/api");

const mockUser = {
  id: 1,
  email: "test@example.com",
  firstName: "Test",
  lastName: "User",
  state: "active",
};

it("renders without crashing", () => {
  render(
    <MemoryRouter>
      <UserContext.Provider value={{ currentUser: mockUser }}>
        <ProfileForm logout={() => {}} />
      </UserContext.Provider>
    </MemoryRouter>
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <UserContext.Provider value={{ currentUser: mockUser }}>
        <ProfileForm logout={() => {}} />
      </UserContext.Provider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("submits form with valid data", async () => {
  userApi.saveProfile.mockResolvedValueOnce({
    ...mockUser,
    firstName: "Updated",
  });

  const { getByLabelText, getByText } = render(
    <MemoryRouter>
      <UserContext.Provider value={{ currentUser: mockUser }}>
        <ProfileForm logout={() => {}} />
      </UserContext.Provider>
    </MemoryRouter>
  );

  fireEvent.change(getByLabelText("First Name"), {
    target: { value: "Updated" },
  });
  fireEvent.click(getByText("Save Changes"));

  await waitFor(() => {
    expect(userApi.saveProfile).toHaveBeenCalledWith(
      1,
      expect.objectContaining({ firstName: "Updated" })
    );
  });
});
