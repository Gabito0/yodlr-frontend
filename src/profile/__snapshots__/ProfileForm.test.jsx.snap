// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container mt-5"
  >
    <div
      class="row justify-content-center"
    >
      <div
        class="col-md-8"
      >
        <div
          class="card"
        >
          <div
            class="card-header"
          >
            <h3>
              Profile
            </h3>
          </div>
          <div
            class="card-body"
          >
            <form>
              <div
                class="form-group"
              >
                <label
                  for="email"
                >
                  Email
                </label>
                <input
                  autocomplete="email"
                  class="form-control"
                  id="email"
                  name="email"
                  type="email"
                  value="test@example.com"
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  for="firstName"
                >
                  First Name
                </label>
                <input
                  class="form-control"
                  id="firstName"
                  name="firstName"
                  value="Test"
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  for="lastName"
                >
                  Last Name
                </label>
                <input
                  class="form-control"
                  id="lastName"
                  name="lastName"
                  value="User"
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  for="currentPassword"
                >
                  Current Password
                </label>
                <input
                  autocomplete="current-password"
                  class="form-control"
                  id="currentPassword"
                  name="currentPassword"
                  type="password"
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  for="newPassword"
                >
                  New Password
                </label>
                <input
                  autocomplete="new-password"
                  class="form-control"
                  id="newPassword"
                  name="newPassword"
                  type="password"
                  value=""
                />
              </div>
              <div
                class="form-group d-flex align-items-center"
              >
                <label
                  class="mr-2"
                  for="state"
                >
                  State:
                </label>
                <p
                  class="form-control-plaintext mb-0 ms-1"
                  id="state"
                >
                  active
                </p>
              </div>
              <div
                class="d-flex justify-content-between"
              >
                <button
                  class="btn btn-primary"
                  type="submit"
                >
                  Save Changes
                </button>
                <button
                  class="btn btn-danger"
                  type="button"
                >
                  Delete Account
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
